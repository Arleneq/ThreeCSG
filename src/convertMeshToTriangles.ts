import {Mesh} from 'three';
import Triangle from './Triangle';
import {isBufferGeometry} from './utils';

export default function convertMeshToTriangles(mesh: Mesh): Triangle[] {
    if (isBufferGeometry(mesh.geometry)) {
        throw new Error(' Only meshes with Three.Geometry are supported.');
    }

    const triangles: Triangle[] = [];

    mesh.updateMatrixWorld(true);
    const { matrixWorld: transform } = mesh;

    const { geometry: { faces, vertices } } = mesh;
    for (let i = 0; i < faces.length; i++) {
        const face = faces[i];

        const a = vertices[face.a].clone();
        const b = vertices[face.b].clone();
        const c = vertices[face.c].clone();
        a.applyMatrix4(transform);
        b.applyMatrix4(transform);
        c.applyMatrix4(transform);

        const triangle = new Triangle(a, b, c);

        triangles.push(triangle);
    }

    return triangles;
}