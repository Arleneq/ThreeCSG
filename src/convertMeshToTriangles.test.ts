import {Mesh, BoxGeometry, Vector3} from 'three';
import convertMeshToTriangles from './convertMeshToTriangles';

describe ('convertMeshToTriangles', () => {
    it('converts a Three.Mesh with non-instanced geometry to Triangles', () => {
        const geometry = new BoxGeometry(5, 8, 8);
        const mesh = new Mesh(geometry);
        mesh.position.set(1, 2, 3);
        mesh.quaternion.set(0, 1, 0, 1).normalize();

        const triangles = convertMeshToTriangles(mesh);
        expect(triangles).toMatchSnapshot();
    });
});