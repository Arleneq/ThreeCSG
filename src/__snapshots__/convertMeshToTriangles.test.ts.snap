// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertMeshToTriangles converts a Three.Mesh with non-instanced geometry to Triangles 1`] = `
Array [
  Triangle {
    "a": Vector3 {
      "x": 5,
      "y": 6,
      "z": 0.5000000000000013,
    },
    "b": Vector3 {
      "x": 5,
      "y": -2,
      "z": 0.5000000000000013,
    },
    "c": Vector3 {
      "x": -2.9999999999999987,
      "y": 6,
      "z": 0.49999999999999956,
    },
    "normal": Vector3 {
      "x": 2.2204460492503136e-16,
      "y": 0,
      "z": -1,
    },
    "w": -0.5000000000000002,
  },
  Triangle {
    "a": Vector3 {
      "x": 5,
      "y": -2,
      "z": 0.5000000000000013,
    },
    "b": Vector3 {
      "x": -2.9999999999999987,
      "y": -2,
      "z": 0.49999999999999956,
    },
    "c": Vector3 {
      "x": -2.9999999999999987,
      "y": 6,
      "z": 0.49999999999999956,
    },
    "normal": Vector3 {
      "x": 2.2204460492503136e-16,
      "y": 0,
      "z": -1,
    },
    "w": -0.5000000000000002,
  },
  Triangle {
    "a": Vector3 {
      "x": -2.9999999999999996,
      "y": 6,
      "z": 5.499999999999998,
    },
    "b": Vector3 {
      "x": -2.9999999999999996,
      "y": -2,
      "z": 5.499999999999998,
    },
    "c": Vector3 {
      "x": 4.999999999999998,
      "y": 6,
      "z": 5.5,
    },
    "normal": Vector3 {
      "x": -2.2204460492503136e-16,
      "y": 0,
      "z": 1,
    },
    "w": 5.499999999999999,
  },
  Triangle {
    "a": Vector3 {
      "x": -2.9999999999999996,
      "y": -2,
      "z": 5.499999999999998,
    },
    "b": Vector3 {
      "x": 4.999999999999998,
      "y": -2,
      "z": 5.5,
    },
    "c": Vector3 {
      "x": 4.999999999999998,
      "y": 6,
      "z": 5.5,
    },
    "normal": Vector3 {
      "x": -2.2204460492503136e-16,
      "y": 0,
      "z": 1,
    },
    "w": 5.499999999999999,
  },
  Triangle {
    "a": Vector3 {
      "x": -2.9999999999999996,
      "y": 6,
      "z": 5.499999999999998,
    },
    "b": Vector3 {
      "x": 4.999999999999998,
      "y": 6,
      "z": 5.5,
    },
    "c": Vector3 {
      "x": -2.9999999999999987,
      "y": 6,
      "z": 0.49999999999999956,
    },
    "normal": Vector3 {
      "x": -0,
      "y": 0.9999999999999999,
      "z": 0,
    },
    "w": 5.999999999999999,
  },
  Triangle {
    "a": Vector3 {
      "x": 4.999999999999998,
      "y": 6,
      "z": 5.5,
    },
    "b": Vector3 {
      "x": 5,
      "y": 6,
      "z": 0.5000000000000013,
    },
    "c": Vector3 {
      "x": -2.9999999999999987,
      "y": 6,
      "z": 0.49999999999999956,
    },
    "normal": Vector3 {
      "x": 0,
      "y": 0.9999999999999999,
      "z": 0,
    },
    "w": 5.999999999999999,
  },
  Triangle {
    "a": Vector3 {
      "x": 4.999999999999998,
      "y": -2,
      "z": 5.5,
    },
    "b": Vector3 {
      "x": -2.9999999999999996,
      "y": -2,
      "z": 5.499999999999998,
    },
    "c": Vector3 {
      "x": 5,
      "y": -2,
      "z": 0.5000000000000013,
    },
    "normal": Vector3 {
      "x": 0,
      "y": -0.9999999999999999,
      "z": -0,
    },
    "w": 1.9999999999999998,
  },
  Triangle {
    "a": Vector3 {
      "x": -2.9999999999999996,
      "y": -2,
      "z": 5.499999999999998,
    },
    "b": Vector3 {
      "x": -2.9999999999999987,
      "y": -2,
      "z": 0.49999999999999956,
    },
    "c": Vector3 {
      "x": 5,
      "y": -2,
      "z": 0.5000000000000013,
    },
    "normal": Vector3 {
      "x": 0,
      "y": -0.9999999999999999,
      "z": 0,
    },
    "w": 1.9999999999999998,
  },
  Triangle {
    "a": Vector3 {
      "x": 4.999999999999998,
      "y": 6,
      "z": 5.5,
    },
    "b": Vector3 {
      "x": 4.999999999999998,
      "y": -2,
      "z": 5.5,
    },
    "c": Vector3 {
      "x": 5,
      "y": 6,
      "z": 0.5000000000000013,
    },
    "normal": Vector3 {
      "x": 1,
      "y": 0,
      "z": 3.552713678800502e-16,
    },
    "w": 5,
  },
  Triangle {
    "a": Vector3 {
      "x": 4.999999999999998,
      "y": -2,
      "z": 5.5,
    },
    "b": Vector3 {
      "x": 5,
      "y": -2,
      "z": 0.5000000000000013,
    },
    "c": Vector3 {
      "x": 5,
      "y": 6,
      "z": 0.5000000000000013,
    },
    "normal": Vector3 {
      "x": 1,
      "y": 0,
      "z": 3.552713678800502e-16,
    },
    "w": 5,
  },
  Triangle {
    "a": Vector3 {
      "x": -2.9999999999999987,
      "y": 6,
      "z": 0.49999999999999956,
    },
    "b": Vector3 {
      "x": -2.9999999999999987,
      "y": -2,
      "z": 0.49999999999999956,
    },
    "c": Vector3 {
      "x": -2.9999999999999996,
      "y": 6,
      "z": 5.499999999999998,
    },
    "normal": Vector3 {
      "x": -1,
      "y": -0,
      "z": -1.776356839400251e-16,
    },
    "w": 2.9999999999999987,
  },
  Triangle {
    "a": Vector3 {
      "x": -2.9999999999999987,
      "y": -2,
      "z": 0.49999999999999956,
    },
    "b": Vector3 {
      "x": -2.9999999999999996,
      "y": -2,
      "z": 5.499999999999998,
    },
    "c": Vector3 {
      "x": -2.9999999999999996,
      "y": 6,
      "z": 5.499999999999998,
    },
    "normal": Vector3 {
      "x": -1,
      "y": 0,
      "z": -1.776356839400251e-16,
    },
    "w": 2.9999999999999987,
  },
]
`;
